File: server/restapi/channel.go
package restapi

import (
	"Chatapp/database"
	"Chatapp/request"
	"Chatapp/response"
	"encoding/json"
	"net/http"
	"strings"

	"github.com/gorilla/mux"
)

func CreateChannel(ctx *Context) {
	var request request.Channel
	_ = json.NewDecoder(ctx.Req.Body).Decode(&request)

	name := strings.TrimSpace(request.Name)
	icon := strings.TrimSpace(request.Icon)
	recipientID := strings.TrimSpace(request.RecipientID)

	if name == "" && recipientID == "" {
		ctx.Res.WriteHeader(http.StatusBadRequest)
		return
	}

	channel, statusCode := database.CreateChannel(name, icon, recipientID, &ctx.User, ctx.Db)
	if statusCode != http.StatusOK {
		ctx.Res.WriteHeader(statusCode)
		return
	}

	recipients := []response.User{}
	if channel.Type == 1 {
		recipient, _ := database.GetUser(recipientID, ctx.Db)
		recipients = append(recipients, response.NewUser(recipient, 0))
	} else {
		recipients = append(recipients, response.NewUser(&ctx.User, 0))
	}

	res_channel := response.NewChannel(channel, recipients)
	ctx.WriteJSON(res_channel)
	ctx.Conn.AddUserToChannel(ctx.User.ID.Hex(), channel.ID.Hex())
}

func GetChannels(ctx *Context) {
	res_channels := response.Channels{}
	channels := database.GetChannels(&ctx.User, ctx.Db)
	for _, channel := range channels {
		recipients := []response.User{}
		for _, recipient := range channel.Recipients {
			if channel.Type == 1 && recipient.Hex() == ctx.User.ID.Hex() {
				continue
			}
			recipient, _ := database.GetUser(recipient.Hex(), ctx.Db)
			recipients = append(recipients, response.NewUser(recipient, 0))
		}

		res_channels = append(res_channels, response.NewChannel(&channel, recipients))
	}

	ctx.WriteJSON(res_channels)
}

func GetChannel(ctx *Context) {
	url_vars := mux.Vars(ctx.Req)
	channel_id := url_vars["id"]

	channel, statusCode := database.GetChannel(channel_id, &ctx.User, ctx.Db)
	if statusCode != http.StatusOK {
		ctx.Res.WriteHeader(statusCode)
		return
	}

	recipients := []response.User{}
	for _, recipient := range channel.Recipients {
		if channel.Type == 1 && recipient.Hex() == ctx.User.ID.Hex() {
			continue
		}
		recipient, _ := database.GetUser(recipient.Hex(), ctx.Db)
		recipients = append(recipients, response.NewUser(recipient, 0))
	}

	res_channel := response.NewChannel(channel, recipients)
	ctx.WriteJSON(res_channel)
}

func EditChannel(ctx *Context) {
	url_vars := mux.Vars(ctx.Req)
	channel_id := url_vars["id"]
	var request request.EditChannel
	_ = json.NewDecoder(ctx.Req.Body).Decode(&request)

	name := strings.TrimSpace(request.Name)
	icon := strings.TrimSpace(request.Icon)

	if name == "" && icon == "" {
		ctx.Res.WriteHeader(http.StatusBadRequest)
		return
	}

	channel, statusCode := database.ModifyChannel(channel_id, name, icon, &ctx.User, ctx.Db)
	if statusCode != http.StatusOK {
		ctx.Res.WriteHeader(statusCode)
		return
	}

	recipients := []response.User{}
	for _, recipient := range channel.Recipients {
		recipient, _ := database.GetUser(recipient.Hex(), ctx.Db)
		recipients = append(recipients, response.NewUser(recipient, 0))
	}

	res_channel := response.NewChannel(channel, recipients)

	ctx.WriteJSON(res_channel)
	ctx.Conn.BroadcastToChannel(res_channel.ID, "CHANNEL_MODIFY", res_channel)
}

func DeleteChannel(ctx *Context) {
	url_vars := mux.Vars(ctx.Req)
	channel_id := url_vars["id"]

	channel, statusCode := database.DeleteChannel(channel_id, &ctx.User, ctx.Db)
	if statusCode != http.StatusOK {
		ctx.Res.WriteHeader(statusCode)
		return
	}

	recipients := []response.User{}
	for _, recipient := range channel.Recipients {
		recipient, _ := database.GetUser(recipient.Hex(), ctx.Db)
		recipients = append(recipients, response.NewUser(recipient, 0))
	}

	res_channel := response.NewChannel(channel, recipients)

	ctx.WriteJSON(res_channel)
	ctx.Conn.RemoveUserFromChannel(ctx.User.ID.Hex(), channel_id)
	ctx.Conn.BroadcastToChannel(channel.ID.Hex(), "CHANNEL_MODIFY", res_channel)
}

File: web/src/api/channel.ts
import Routes from "../config";
import { ChannelOBJ } from "../models/models";

export async function GetChannels() {
    const response = await fetch(Routes.Channels);
    if (!response.ok) {
        return [] as ChannelOBJ[];
    }

    const channels: ChannelOBJ[] = await response.json();
    return channels;
}

export async function CreateChannel(name: string, icon: string | ArrayBuffer | null) {
    const url = Routes.currentUser+"/channels";
    const response = await fetch(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({ name: name, icon: icon })
    })

    if (!response.ok) {
        return {} as ChannelOBJ;
    }

    const channel_: ChannelOBJ = await response.json();
    return channel_;
}

export async function EditChannel(channel_id: string, name: string, icon: string) {
    const url = Routes.Channels+"/"+channel_id;
    const response = await fetch(url, {
        method: "PATCH",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({ name: name, icon: icon })
    })
    return response;
}

export async function DeleteChannel(channel_id: string) {
    const url = Routes.Channels+"/"+channel_id;
    const response = await fetch(url, {
        method: "DELETE",
    })
    return response;
}

export async function GetDMChannel(user_id: string) {
    const url = Routes.currentUser+"/channels"
    const response = await fetch(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({ recipient_id: user_id })
    })
    return response;
}